pallets/account/src/lib.rs

```diff
@@ -0,0 +1,308 @@
+        type ClaimSignature: ClaimSignature<AccountId = Self::AccountId, Address = EvmAddress>;
+
+        // /// Weight information for the extrinsics in this module
+        // type WeightInfo: WeightInfo;
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// AccountId has mapped
+        AccountIdHasMapped,
+        /// Eth address has mapped
+        EthAddressHasMapped,
+        /// Bad signature
+        BadSignature,
+        /// Invalid signature
+        InvalidSignature,
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// EVM Account claimed.
+        /// Double Mapping b/w native and evm address created
+        AccountClaimed {
+            account_id: T::AccountId,
+            evm_address: EvmAddress,
+        },
+    }
+
+    /// Native accounts for EVM address
+    /// NativeAccounts: EvmAddress => Option<AccountId>
+    #[pallet::storage]
+    pub type NativeAccounts<T: Config> =
+        StorageMap<_, Twox64Concat, EvmAddress, T::AccountId, OptionQuery>;
+
+    /// EVM Addresses for native accounts
+    /// EvmAccounts: AccountId => Option<EvmAddress>
+    #[pallet::storage]
+    pub type EvmAccounts<T: Config> =
+        StorageMap<_, Twox64Concat, T::AccountId, EvmAddress, OptionQuery>;
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Claim account mapping between Substrate accounts and EVM accounts.
+        /// Ensure no prior mapping exists for evm address.
+        ///
+        /// - `evm_address`: The address to bind to the caller's account
+        /// - `signature`: A signature generated by the address to prove ownership
+        #[pallet::call_index(0)]
+        #[pallet::weight(0)]
+        pub fn claim_evm_account(
+            origin: OriginFor<T>,
+            evm_address: EvmAddress,
+            signature: SignatureOf<T>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            // make sure no prior mapping exists
+            Self::enure_no_mapping(&who, &Some(evm_address))?;
+            // claim the address
+            Self::do_claim_address(who, evm_address, signature)
+        }
+
+        /// Claim default evm address for given account id
+        /// Ensure no prior mapping exists for the account
+        #[pallet::call_index(1)]
+        pub fn claim_default_evm_account(origin: OriginFor<T>) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            // make sure no prior mapping exists
+            Self::enure_no_mapping(&who, &None)?;
+            // claim default address
+            let _ = Self::do_claim_default_address(who)?;
+            Ok(())
+        }
+    }
+}
+
+impl<T: Config> Pallet<T> {
+    /// Ensure no mappings exists for given pair of account/address
+    fn enure_no_mapping(account_id: &T::AccountId, address: &Option<EvmAddress>) -> DispatchResult {
```

Comment:
Good point, will do that. Although it is very unlikely they can collide but no harm in making sure they won't

Reply:
Yes, that's true and that cannot be prevented since it's the same thing as finding the private key.

But at least storage will stay in consistent state, unlike as with the scenario described above.

Reply:
> If an EVM address is being claimed, be it a default one or a custom one, you need to ensure it's not already in use.

This only applies for default address only, since custom (EOA) account most likely be already used by user before connecting them with native.

Reply:
I'll add the check in default claim dispatch to make sure generated evm address is not already mapped